google/cloud/pubsub/internal/extend_leases_with_retry_test.cc

```diff
@@ -0,0 +1,175 @@
+          make_status_or(std::chrono::system_clock::now())))));
+  EXPECT_CALL(*mock, AsyncModifyAckDeadline(
+                         _, _,
+                         Property(&ModifyAckDeadlineRequest::ack_ids,
+                                  ElementsAre("test-001", "test-002"))))
+      .WillOnce(Return(ByMove(make_ready_future(Status{}))));
+
+  ModifyAckDeadlineRequest request;
+  request.add_ack_ids("test-001");
+  request.add_ack_ids("test-002");
+  auto result = ExtendLeasesWithRetry(mock, CompletionQueue(mock_cq), request);
+  EXPECT_STATUS_OK(result.get());
+}
+
+TEST(ExtendLeasesWithRetry, SuccessWithPartials) {
+  auto mock = std::make_shared<MockSubscriberStub>();
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+
+  ::testing::InSequence sequence;
+  EXPECT_CALL(*mock, AsyncModifyAckDeadline(
+                         _, _,
+                         Property(&ModifyAckDeadlineRequest::ack_ids,
+                                  ElementsAre("test-001", "test-002",
+                                              "test-003", "test-004"))))
+      .WillOnce(Return(ByMove(make_ready_future(MakeStatusWithDetails({
+          {"test-001", "TRANSIENT_FAILURE_1"},
+          {"test-002", "TRANSIENT_FAILURE_2"},
+          {"test-003", "PERMANENT_BADNESS"},
+      })))));
+  EXPECT_CALL(*mock_cq, MakeRelativeTimer)
+      .WillOnce(Return(ByMove(make_ready_future(
+          make_status_or(std::chrono::system_clock::now())))));
+  EXPECT_CALL(*mock, AsyncModifyAckDeadline(
+                         _, _,
+                         Property(&ModifyAckDeadlineRequest::ack_ids,
+                                  ElementsAre("test-001", "test-002"))))
+      .WillOnce(Return(ByMove(make_ready_future(Status{}))));
+
+  ModifyAckDeadlineRequest request;
+  request.add_ack_ids("test-001");
+  request.add_ack_ids("test-002");
+  request.add_ack_ids("test-003");
+  request.add_ack_ids("test-004");
+  auto result = ExtendLeasesWithRetry(mock, CompletionQueue(mock_cq), request);
+  EXPECT_STATUS_OK(result.get());
+}
+
+TEST(ExtendLeasesWithRetry, FailurePermanentError) {
+  auto mock = std::make_shared<MockSubscriberStub>();
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+
+  EXPECT_CALL(*mock, AsyncModifyAckDeadline(
+                         _, _,
+                         Property(&ModifyAckDeadlineRequest::ack_ids,
+                                  ElementsAre("test-001", "test-002"))))
+      .WillOnce(Return(ByMove(make_ready_future(MakeStatusWithDetails({})))));
+
+  ModifyAckDeadlineRequest request;
+  request.add_ack_ids("test-001");
+  request.add_ack_ids("test-002");
+  auto result = ExtendLeasesWithRetry(mock, CompletionQueue(mock_cq), request);
+  EXPECT_THAT(result.get(), StatusIs(StatusCode::kUnknown));
+}
+
+TEST(ExtendLeasesWithRetry, FailureTooManyTransients) {
+  auto mock = std::make_shared<MockSubscriberStub>();
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+
+  auto timer = [] {
```

Comment:
nit: is there any reason why `timer` and `transient` are defined out here? it is inconsistent with the rest of the tests

Reply:
When using `WillRepeatedly()` you cannot use `ByMove()`, but now that we are using `WillOnce()` this is not needed